enhancements/ebpf-in-ocp.md

```diff
@@ -0,0 +1,211 @@
+`MachineConfig` (or something similar) to deploy a small (or
+not-so-small) eBPF program to every node in my cluster, without
+voiding my warranty.
+
+I might be OK with needing a support exception.
+
+(eg, concretely, there is a customer who wants to use eBPF to
+implement [OVS-style SLB
+Bonding](https://docs.openvswitch.org/en/latest/topics/bonding/#slb-bonding)
+on a kernel bond interface, using eBPF. The upstream Linux maintainer
+rejected a patch to implement it in the kernel driver directly, in
+part because it would be so easy to implement it with eBPF.)
+
+#### Cilium as a Supported Third-Party Network Plugin
+
+As an OCP customer, I want to deploy an OCP cluster using Cilium as a
+certified third-party network plugin, without worrying that any other
+OCP components are going to interfere with it (or vice versa).
+
+Cilium uses a wide variety of eBPF hooks, including FIXME.
+
+#### eBPF From OLM
+
+As a third-party developer of an operator that uses eBPF, I would like
+to make my operator available via OLM and have it interoperate
+smoothly with existing and future OCP releases.
+
+#### Future eBPF Data Plane Functionality
+
+As an OCP developer/product manager, I would like to introduce
+functionality in a future release that makes use of the XDP and/or tc
+eBPF hooks to modify data plane functionality, and have these new
+features interact in a predictable way with existing networking
+functionality, other operators, and local customizations.
+
+#### Security
+
+As a cluster administrator, I would like to make sure that no one is
+using eBPF that I don't know about, because I am in a highly regulated
+industry, and I can't have random people running mystery code in my
+kernel, even if it's sandboxed.
+
+### Goals
+
+- Categorize the existing, near-term-planned, and long-term-envisioned
+  uses of eBPF in OCP.
+
+- Create a set of rules for using eBPF in OCP to ensure that different
+  eBPF users do not clobber each other's programs and do not cause
+  system-wide instability or un-debuggability.
+
+  - We may need different rules for different kinds of eBPF
+    programs.
+
+### Non-Goals
+
+-
+
+## Proposal
+
+TBD
+
+### Workflow Description
+
+N/A - not an end-user feature
+
+### API Extensions
+
+TBD - none expected
+
+### Implementation Details/Notes/Constraints [optional]
+
+### Risks and Mitigations
+
+- Continuing to allow arbitrary eBPF without any oversight will likely
+  eventually lead to buggy clusters, difficult-to-debug problems, and
+  possibly security holes.
+
+- OTOH adopting a too-strict policy toward customer use of eBPF may
+  turn some customers away (and would be awkward, messaging-wise:
+  "eBPF is great! You can't use it!").
+
+  - We should look into what RHEL is doing to balance
+    "supportability" vs "letting the customer do what they want"
+    with eBPF.
+
+### Drawbacks
+
+None. We really need guidelines here.
+
+## Design Details
+
+### Open Questions [optional]
+
+Many
```

Comment:
Open question:

Where do we stand with support on RHEL. The [latest release notes](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/9/html/9.0_release_notes/new-features) state:

>  The XDP and AF_XDP API parts of the libbpf library are not supported and may be removed in a future release. 

We may not use `libbpf` to load XDP or AF_XDP programs.

> The eXpress Data Path (XDP) feature, which provides access to the received packets before the kernel networking stack processes them. Red Hat supports XDP only if it is used through the libxdp library.

We must use libxdp BUT we do not have higher level language bindings for it (i.e for Go).

> The xdp-tools package, which contains user-space support utilities for the XDP feature and is supported on the AMD64 and Intel64 CPU architectures.

libxdp is not packaged for any architecture other than x86_64

Reply:
For RHEL9, libxdp is packaged for all archs, it just won't do multi-prog until the kernel support lands; but that's fine, it'll fall back automatically, so it should be totally fine to use it as-is for those archs. So the lack of Go bindings is the only blocker AFAICT...

Reply:
thanks for clarifying @tohojo
we'll have to figure out a design for libxdp-go that works seamlessly with `cillium/ebpf`
lets' chat about that next week and perhaps we can get something started